Make verify-typecheck not depend on GOPATH

thockin opened the pull request:
This is a total rewrite of test/typecheck against go/packages.  It now runs under GOPATH or not.

CAVEAT: the way I make it work outside of GOPATH is by faking GOPATH.  When it thinks it is in modules mode, it loses its mind because we have other go.mod files under our tree and the tooling simply can not handle that.

To be really module-friendly requires more work.

/kind bug

xref #92253

```release-note
NONE
```

thockin committed:
Make verify-typecheck not depend on GOPATH

This script fails if not under a GOPATH because we use go/types, which
is not modules aware.

thockin committed:
Don't typecheck directories that start with _

Go ignores them, so should we.

liggitt commented on test/typecheck/main.go:
```diff
@@ -266,7 +266,7 @@ func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
 	if info.IsDir() {
 		// Ignore hidden directories (.git, .cache, etc)
-		if len(path) > 1 && path[0] == '.' {
+		if len(path) > 1 && (path[0] == '.' || path[0] == '_') {
```
is this related to GOPATH or something else?

thockin replied:
similar but different - I happen to have an _thockin dir with little tools I occasionally use, including some Go code.  Tools should all ignore that.  Typecheck did not ignore it.

liggitt replied:
This will ignore things like https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/_examples as well, right? I guess that's a feature, not a bug.

thockin commented:
Yes.  If Go ignores it, we should ignore it.  IMO.

On Tue, Jun 16, 2020 at 10:59 PM Jordan Liggitt <notifications@github.com>
wrote:

> *@liggitt* commented on this pull request.
> ------------------------------
>
> In test/typecheck/main.go
> <https://github.com/kubernetes/kubernetes/pull/92212#discussion_r441298247>
> :
>
> > @@ -266,7 +266,7 @@ func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
>  	}
>  	if info.IsDir() {
>  		// Ignore hidden directories (.git, .cache, etc)
> -		if len(path) > 1 && path[0] == '.' {
> +		if len(path) > 1 && (path[0] == '.' || path[0] == '_') {
>
> This will ignore things like
> https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/_examples
> as well, right? I guess that's a feature, not a bug.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/92212#discussion_r441298247>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBWGJNYSFFMBZQKJYDRXBLUHANCNFSM4OAIBQQA>
> .
>

liggitt commented:
/lgtm
/approve

nikhita commented:
/hold
that's a legit failure - https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92212/pull-kubernetes-typecheck/1273127809723666434

I think this occurs because `go/types` doesn't work well with modules. The recommendation from the go team is to use [`go/packages`](https://godoc.org/golang.org/x/tools/go/packages) going forward for type checking in module-friendly mode. Ref: https://github.com/golang/go/issues/27556#issuecomment-419468978, https://github.com/golang/go/issues/28328

thockin commented:
@nikhita Weird - it runs locally.

The alternative id to force GOPATH via `run-in-gopath.sh`.  I am disinclined to rewrite a tool in a totally new library.  Yay, Go.

thockin commented:
Trying the run-in-gopath approach instead

thockin commented:
/retest

thockin commented:
/retest

nikhita commented on hack/verify-typecheck.sh:
```diff
@@ -30,8 +30,13 @@ cd "${KUBE_ROOT}"
+# As of June, 2020 the typecheck tool is written in terms of go/types, but that
+# library doesn't work well with modules.  Guidance is to rewrite tools against
+# golang.org/x/tools/go/packages.  Until that is done, force this tooling to
+# run in a fake GOPATH.
```
Created https://github.com/kubernetes/kubernetes/issues/92253 to track this.

Will let @liggitt lgtm when he's back.

thockin replied:
Pushing a fix, but with caveats.  Will update the PR comment here.  PTAL as if from scratch

thockin commented:
* removing the hold
* looking for LGTM
* working on a better version of this in the background, but go modules is a mess when you cross boundaries.

nikhita commented:
lgtm'ing since we are tracking efforts to rewrite the typecheck tool in https://github.com/kubernetes/kubernetes/issues/92253 :+1: 

/lgtm

thockin committed:
Rewrite typecheck against x/tools/go/packages

This runs much faster than before.  This change removes all of the
async status output because all of the compute time is spent inside
go/packages, with no opportunity to update the status.

Adds testdata code to prove it fails when expected.

thockin commented:
Please take a new look.  It's a near-total rewrite now.

BenTheElder commented:
/retest

SergeyKanzhelev commented on hack/verify-typecheck.sh:
```diff
@@ -30,8 +30,12 @@ cd "${KUBE_ROOT}"
 make --no-print-directory -C "${KUBE_ROOT}" generated_files
 
+# As of June, 2020 the typecheck tool is written in terms of go/packages, but
+# that library doesn't work well with multiple modules.  Until that is done,
```
it's not clear in this sentence until what is done.

nikhita replied:
Yeah, I'm a little confused why `run-in-gopath` in still needed.

thockin replied:
Our tree has multiple `go.mod` in it.  We're making a tool that does filesystem walk, which ignores go.mod.  These are not compatible,

We have to choose - we can ignore go.mod, in which case it has to be under a GOPATH or we can respect go.mod which requires that we run this tool multiple times against each module (or something, TBD).

I chose to rewrite to go/packages first.  Hence GOPATH

SergeyKanzhelev replied:
I'd suggest to expand the comment a little bit. When reading it out of context it's unclear what this `Until that` refers to =). Maybe "Until it works well with multiple modules,"

BenTheElder commented:
One follow-up / cleanup note: https://github.com/kubernetes/kubernetes/commit/2682f309ccb694c1cbf8b4428d639902f2096b52#r40021683

/lgtm
/approve

BenTheElder commented:
@thockin its your favorite: needs a hack/update-bazel.sh by the looks of the failures...

nikhita commented:
/lgtm cancel

these are legit failures :)

thockin commented:
There's a lot of overlap between typecheck and staticcheck.  We should probably only keep one, and staticcheck seems more comprehensive.

thockin commented:
Updated to include vendor/ properly.  Had to special-case CGo errors.

I can't get staticcheck to run on my desktop now, but the errors is lists in the test fails indicate a high degree of overlap with typecheck.  Maybe we should just run staticcheck against each architecture and get rid of typecheck?

thockin commented:
false start - pushed without Bazel files

BenTheElder commented:
> I can't get staticcheck to run on my desktop now, but the errors it lists in the test fails indicate a high degree of overlap with typecheck. Maybe we should just run staticcheck against each architecture and get rid of typecheck?

typecheck being relatively fast has helped us cover more platforms, but also go tags on/off (e.g. `dockerless` == no docker shim and `providerless` == attempt to compile out in-tree cloud providers).

If we can do this with staticcheck, sure.
If not for the whole go module thing, we wouldn't have really needed to touch this code again though, it's "just worked".

thockin commented:
go modules infects everything it touches.  And it didn't even live up to my hopes.

I still need to have my src under a GOPATH for some tools.
I still can't have symlinks in the path to my src.
Multiple modules in a tree is a disaster for tools.

thockin commented:
How do I tell bazel to not try to compile code in a `testdata` directory?  Sigh.

BenTheElder commented:
seems gazelle still requires explicit instruction to ignore a testdata dir...

> You can add a comment # gazelle:exclude testdata to the top of the build file in the parent directory. 

https://github.com/bazelbuild/rules_go/issues/415#issuecomment-348230264

also needs a rebase now ... staging has multiplied.

thockin committed:
Add a reminder to add/rm files

thockin commented:
@BenTheElder I added the gazelle comment, but `test/typecheck/BUILD` still shows:

```
filegroup(
    name = "all-srcs",
    srcs = [
        ":package-srcs",
        "//test/typecheck/testdata/good/testdata:all-srcs",
    ],
    tags = ["automanaged"],
)
```

Remind me again why we keep Bazel?  How do I tell it "no really, ignore testdata" ?

BenTheElder commented on test/typecheck/BUILD:
```diff
@@ -44,4 +45,5 @@ go_test(
     name = "go_default_test",
     srcs = ["main_test.go"],
     embed = [":go_default_library"],
+    deps = ["//vendor/golang.org/x/tools/go/packages:go_default_library"],
```
this needs a `data` entry with the testdata, I think roughly `data = ["//test/typecheck/testdata:all-srcs"]` though that target doesn't appear to exist ...

BenTheElder commented on test/typecheck/testdata/good/_ignore/BUILD:
```diff
@@ -0,0 +1,14 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
+
+go_library(
```
agh, this shouldn't exist, maybe we need stub BUILD files above this with the gazelle comment?

BenTheElder commented:
I haven't been in favor of keeping bazel myself for a long time now but as I understand it the most recent discussion on that was "above my paygrade" ... :upside_down_face: 

---

I think kazel generates these "all-srcs" rules, so that part is on kubernetes ...
https://github.com/kubernetes/repo-infra/tree/master/cmd/kazel

That shouldn't be a problem I think, we actually need something like that to pass as a data dependency to the go_test rule.

thockin commented:
Oh FFS.  I wrote the tests and now they can't run because they depend on `go` being in the PATH.

@BenTheElder - opinions on whether this is something that will realistically be fixable, or should I just delete the test?

```
=== RUN   TestVerify
    TestVerify: main_test.go:45: unexpected error: 'go list' driver requires 'go', but executable file not found in $PATH
    TestVerify: main_test.go:45: unexpected error: 'go list' driver requires 'go', but executable file not found in $PATH
```

BenTheElder commented:
sigh https://github.com/bazelbuild/rules_go/issues/1996

this PR now needs rebase from the go1.14. bump.

fejta commented:
> they depend on go being in the PATH.

If the test wants to use the go binary then it should declare its dependnecy on it, which you can do via:

```py
args = ["--go=$(location @go_sdk//:bin/go"]
data = ["@go_sdk//:bin/go"]
```

For example: https://github.com/kubernetes/repo-infra/blob/c570ba2f47280cd25c31c9f26ffdb6a17330c109/hack/BUILD.bazel#L109

This will also ensure that it uses the correct version of golang as opposed to whatever random version is installed on the path

thockin commented:
ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/test/typecheck/BUILD:41:1: in args attribute of go_test rule //test/typecheck:go_default_test: label '@go_sdk//:bin' in $(location) expression is not a declared prerequisite of this rule. Since this rule was created by the macro 'go_test_macro', the error might have been caused by the macro implementation

thockin commented:
Now it passes in a relative path that is relative to some root, but that root is NOT the CWD of the test binary.

I am sick and tired of Bazel.  How much money do we spend on test infra running and re-running CI against PRs like this, that are just repeatedly pushing, trying to beat bazel into submission?  This PR alone has probably cost north of $100.

thockin commented:
@fejta I'm going to need more clues.

I pass in the arg, as you suggested (thanks) and I get a relative path: "external/go_sdk/bin/go".  The problem is that it is not relative to my CWD, it is relative to some other directory a few levels up from me.  I can not find any Bazel magic that tells me what that root directory is.  As such, this is not working.  

https://docs.bazel.build/versions/master/be/make-variables.html suggests execpath or rootpath, but empirically they produce the same result as $(location).  It also suggests RULEDIR which simply fails to run (not defined).

I am at a loss.  I am a hair away from just abandoning this change.  Can you offer any guidance?  I can't find anything in docs, but who knows if I am searching in the right places.

Relevant lines (trimmed a bit)
```
HACK: goBinary = "external/go_sdk/bin/go"
HACK:  <nil> stat external/go_sdk/bin/go: no such file or directory
HACK: CWD:  /b/f/w/bazel-out/k8-fastbuild/bin/test/typecheck/linux_amd64_race_stripped/go_default_test.runfiles/io_k8s_kubernetes/test/typecheck
find: ../../../../go_default_test.runfiles/go_sdk/bin/go
find: ../../../../go_default_test.runfiles/io_k8s_kubernetes/external/go_sdk/bin/go

thockin commented:
I hacked a hardcode "../.." in. Puke.

thockin commented:
/retest pull-kubernetes-e2e-kind-ipv6

thockin commented:
/test pull-kubernetes-e2e-kind-ipv6

BenTheElder commented:
IPv6 job is non-blocking / not `Required` and shouldn't be affected by these changes, and I think currently struggling. @aojea is investigating.

BenTheElder commented:
or in this case it's just that racy test case we were talking about again @aojea 
```
[sig-cli] Kubectl client Simple pod should support inline execution and attach expand_less | 1m2s

test/e2e/kubectl/kubectl.go:549 Jun 26 18:35:17.769: Expected     <string>:  to contain substring     <string>: stdin closed test/e2e/kubectl/kubectl.go:573
```

thockin commented:
Holy carp, it's green, someone say LGTM, quick!!

BenTheElder commented:
/lgtm
/approve

BenTheElder commented:
/retest

BenTheElder commented:
> pull-kubernetes-e2e-gce-ubuntu-containerd — Job failed.

https://github.com/kubernetes/kubernetes/issues/92563
/retest
